### PR TITLE
feat(core): add vue2 support

### PR DESCRIPTION
**Pull request by Toilal**
I try to include vue2 support to this module. This is currently a draft because it seems to compile properly but I still need to do some tests.

### CODE REVIEW

**Comment by FranckFreiburger** [comment]
Interesting !
Just keep in mind that my first concern is the final bundle size.

**Comment by Toilal** [comment]
Yes, I just saw the dist file is huge now in this branch, but there's probably some tweeks to perform in webpack configuration. Maybe it should build another file for vue2 (with another webpack entrypoint).

**Comment by FranckFreiburger** [comment]
Hi,

FYI, vue3-sfc-loader src has been refactored.
`src/createSFCModule.ts` is now an "interface". see webpack config :

```
alias: {
	'./createSFCModule.ts': `./createVue${ vueVersion }SFCModule.ts`,
	...
```

**Comment by Toilal** [comment]
Thanks. 

The webpack alias is an option, but it seems to me there's mostly common behaviors to share between the `createVue2SFCModule.ts` and `createVue3SFCModule.ts`, so this will lead to some duplicated code (even if you have extracted some common functions in tools.js, there's still many features that needs to be duplicated like componentHash/scopeId, language processor preloading, global workflow, ...).

I currently have to build an app where end-user can build new forms through a back office UI. My idea is to build a WYSIWYG editor supporting some vuetify components through live compiling of SFC files, but vuetify still doesn't support vue3 (it's just been released in alpha though), so I currently need vue2 support. 

I also have some side project ideas using this "live compiling Vue SFC stuff", so I still hesitate into contributing to your repository or creating my own set of libraries. In such a case, it would be a monorepo with many published packages (compiler-core, compiler-vue2, compiler-vue3, babel-vue3, ...). I wish to have a node version too, using node internal modules instead of browserify ones. But anyway for now i'll try to rebase the vue2 version with your changes.

**Comment by Toilal** [comment]
Ok, I have rework the vue2 integration with your changes. It seems OK to me, the first "Hello World" test in OK in `basic.vue2.test.js` and i'll try to port all vue3 tests.

**Review by FranckFreiburger — APPROVED** [review]

File: build/webpack.config.js [hunk 1]
```diff
@@ -157,8 +158,16 @@ ${ pkg.name } v${ pkg.version }
...
 			`.trim()),
 		],
-
+		externals: {
+			'vue': {
+				amd: 'vue',
+				commonjs: 'vue',
+				commonjs2: 'vue',
+				root: 'Vue'
+			},
```
Comment (FranckFreiburger) [inline build/webpack.config.js:167]: Just wondering why it is needed ?

File: package.json [hunk 2]
```diff
@@ -62,10 +63,10 @@
     "stream-browserify": "^3.0.0",
     "url": "^0.11.0",
     "util": "^0.12.3",
-    "vm-browserify": "^1.1.2",
-    "vue": "3.0.4"
+    "vm-browserify": "^1.1.2"
   },
   "devDependencies": {
+    "@vue/component-compiler-utils": "^3.2.0",
```
Comment (FranckFreiburger) [inline package.json:69]: since it's part of the final package, move to "dependencies" ?

File: package.json [hunk 3]
```diff
@@ -79,8 +80,16 @@
     "typedoc": "0.19.2",
     "typedoc-plugin-markdown": "3.1.1",
     "typescript": "^4.2.2",
+    "vue": "^3.0.7",
+    "vue-template-compiler": "^2.6.12",
```
Comment (FranckFreiburger) [inline package.json:84]: to "dependencies" ?

File: package.json [hunk 4]
```diff
@@ -79,8 +80,16 @@
...
+    "vue2": "npm:vue@^2.6.12",
     "webpack": "^5.24.3",
     "webpack-bundle-analyzer": "^4",
     "webpack-cli": "^4.5.0"
+  },
+  "peerDependencies": {
+    "@vue/component-compiler-utils": "^3.2.0",
+    "vue": "^2.6.12|^3.0.7",
+    "vue-template-compiler": "^2.6.12"
```
Comment (FranckFreiburger) [inline package.json:93]: why peerDependencies since it's already in normal dependencies ?

File: src/createVue2SFCModule.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	const compileTemplateOptions : TemplateCompileOptions = descriptor.template ? {
+		// hack, since sourceMap is not configurable an we want to get rid of source-map dependency. see genSourcemap
+		source: descriptor.template.content,
+		filename,
+		compiler: vueTemplateCompiler as VueTemplateCompiler,
+		compilerOptions: undefined,
+		preprocessLang: descriptor.template.lang,
+		isProduction: isProd,
+		prettify: false
+	} : null;
```
Comment (FranckFreiburger) [inline src/createVue2SFCModule.ts:85]: no scopeId here, should we handle scope attribute manually ?

File: package.json [hunk 6]
```diff
@@ -79,8 +80,16 @@
     "typedoc": "0.19.2",
     "typedoc-plugin-markdown": "3.1.1",
     "typescript": "^4.2.2",
+    "vue": "^3.0.7",
```
Comment (FranckFreiburger) [inline package.json:83]: maybe "vue3": "npm:vue@^3.0.7", ?

**Comment by FranckFreiburger** [comment]
looks really promising (happy to be able to archive http-vue-loader soon).
About createVue2SFCModule.ts and createVue3SFCModule.ts factorization, I agree that we can do more, maybe a bit later ?

btw, thanks for your contribution !

PS: will your WYSIWYG view editor be released publicly ?

File: build/webpack.config.js (hunk 1 above)
Comment (Toilal) [reply build/webpack.config.js:167]: I'm not sure it's needed.

File: src/createVue2SFCModule.ts (hunk 5 above)
Comment (Toilal) [reply src/createVue2SFCModule.ts:85]: Yes, this is still a thing that doesn't work with vue2. It seems `vue-template-compiler` doesn't support the addition of the scope id data attribute on template root element. We need to find a way to add it manually to the component element. But it's not that simple as there's no reference to the componentDOM element inside `getStyle` callback. Maybe we should hack the generated `render` function ?

File: package.json (hunk 6 above)
Comment (Toilal) [reply package.json:83]: As you wish. I would let the current vue version the default one (default alias = vue).

**Comment by Toilal** [comment]
Thanks for merging this pull request. Of course, it can be refactored later. For now, i'll try to make scoped style work properly and copy/paste tests from vue3 version.

I wish to release a WYSIWYG view editor publicly, but it won't happen shortly.

**Pull request merged**